core/src/ast/methods.rs

```diff
@@ -115,69 +114,100 @@ impl Method {
     pub fn params_held_by_output(&self) -> (Option<&SelfParam>, Vec<&Param>) {
         // To determine which params the return type is bound to, we just have to
         // find the params that contain a lifetime that's also in the return type.
-        self.return_type
-            .as_ref()
-            .map(|return_type| {
-                // Collect all lifetimes from return type into a `BTreeSet`.
-                let mut lifetimes = BTreeSet::new();
-                return_type.visit_lifetimes(&mut |lt| -> ControlFlow<()> {
-                    match lt {
-                        Lifetime::Named(name) => {
-                            lifetimes.insert(name);
-                        }
+        if let Some(ref return_type) = self.return_type {
+            // Collect all lifetimes that the return type might care about,
+            // including the lifetimes it contains, as well as the lifetimes
+            // that must live as long as the lifetimes it contains.
+            let lifetimes = {
+                let mut stack = vec![];
+
+                // Populate the initial lifetimes
+                return_type.visit_lifetimes(&mut |lifetime| -> ControlFlow<()> {
+                    match lifetime {
+                        Lifetime::Named(named) => stack.push(named),
                         Lifetime::Anonymous => {
                             // This is also caught in the validity check, so this should
                             // never happen. If we're guaranteed to always run the
                             // check first, we can change this to `unreachable!()`.
                             panic!("Anonymous lifetimes not yet allowed in return types")
                         }
-                        Lifetime::Static => {}
+                        Lifetime::Static => {
+                            // If the output depends on the static lifetime, this
+                            // tells us nothing about what params it might need,
+                            // so do nothing.
+                        }
                     }
                     ControlFlow::Continue(())
                 });
 
-                let held_self_param = self.self_param.as_ref().filter(|self_param| {
-                    // Check if `self` is a reference with a lifetime in the return type.
-                    if let Some((Lifetime::Named(ref name), _)) = self_param.reference {
-                        if lifetimes.contains(name) {
-                            return true;
-                        }
+                let mut lifetimes = Vec::with_capacity(stack.len());
+
+                let mut unused_rules: Vec<&LifetimeDef> =
+                    self.introduced_lifetimes.iter().collect();
+
+                while let Some(sublifetime) = stack.pop() {
```

Comment:
nit: this code looks right, but should have a comment explaining what it's doing (and why)